app/controllers/citizens_controller.rb

```diff
@@ -0,0 +1,54 @@
+class CitizensController < ApplicationController
+  before_action :find_citizen, except: %i[create index]
+
+	# GET /citizens
+	def index
+		@finalResult = []
+		@citizens = Citizen.all
+		@citizens.each do |citizen|
+			@finalResult.push(citizen.json)
+		end
+		render json: @finalResult, status: :ok
+	end
+
+	# GET /citizen/{id}
+	def show
+		render json: @citizen, status: :ok
+	end
+
+	# POST /citizen
+	def create
+    
+		@citizen = Citizen.new(citizen_params)
+		if @citizen.save
+
+			render json: @citizen, status: :created
+		else
+			render json: { errors: @citizen.errors.full_messages }, status: :unprocessable_entity
+		end
+	end
+
+	# PUT /citizen/{id}
+	def update
+		if @citizen.update(citizen_params)
+			render json: @citizen, status: :ok
+		else
+			render json: { errors: @citizen.errors.full_messages }, status: :unprocessable_entity
+		end
+	end
+
+	private
+
+	def find_citizen
+		@citizen = Citizen.find(params[:_id])
+		rescue ActiveRecord::RecordNotFound
+			render json: { errors: 'Citizen not found' }, status: :not_found
```

Comment:
Coloca `I18n` nessas mensagens de erro, caso precise exibir em tela fica mais fácil e é o padrão

Reply:
Pode deixar

Reply:
Remover o render nesse caso depende do propósito da aplicação ou do time de desenvolvimento/produto.

1. Se é um caso onde o produto entende que DEVE DAR ERRO SE NAO OCORRER COMO DEVERIA, o certo é deixar como `find`, e,  a exception será levantada, nesse caso acontecendo um 500 (internal server error). Em rails apps monolito isso acaba sendo tratado e um crud default pode retornar 404 (mas a exception será levantada de qualquer forma).

2. No caso de resource not found (404), quase sempre queremos tratar o erro. Usar rescue é correto também, porém, o código que mostrei é mais otimizado em termos de memória além de flexibilizar o http status.

3. Há uma diferença entre retornar 500 e 404, tanto a back quanto front. No entendimento geral, 404 só não foi encontrado, enquanto 500 quer dizer que algo que não deveria acontecer, aconteceu. Um exemplo disso é no frontend SPA onde o comportamento dos componentes podem mudar de acordo com o http status (`.then((response) =>  setState(response.data)).catch(error => throw Error(error))` sendo um caso rápido de onde o http status IMPORTA.

Resumindo, fazendo um if com find_by e retornando o json com http status not_found é simples e eficiente, na minha opinião.

Reply:
@ricardopacheco falou bonito, mas o que a gente quis dizer é que o metodo `find_citizen` como diz o nome só deveria se preocupar com encontrar ou não o municipe e retornar, o tratamento disso em conjunto com o que renderizar ficam mais entendíveis e mais corretos dentro das actions do controller que é onde esse tipo de fluxo é justamente controlado.

Reply:
Vejam a linha 2, temos um before_action, se perguntem, por que adicionar um `render` dentro de um metodo que usa `before_action`?

Reply:
É o seguinte o metodo render está ai pelo fato de esse metodo, ser usado para setar um cidadão por id no contexto dos metodos show e update, caso ele encontre show de bola, ele passa e continua o processo, caso ele não encontre, quero poder retornar ao front uma mensagem orientando o que aconteceu, se ele achou ele apresenta o show ou completa o update normal, caso ele não encontre nos dois casos eu preciso apresentar uma mensagem de retorno para orientar que o motivo não fo um erro e sim a não identificação de tal id, como a mensagem seria a mesma, faz sentido ela ser mostrada or esta função, ou seja bateu nela achou segue, não achou idenpendente se show ou update ele já renderiza a mensagem, removendo ele dessa função eu teria que tratar o mesmo erro duas vezes uma para metodo show e uma para o metodo update, ao meu ver seria repetivo e desnecessario, pois esta forma soluciona o problema resove tudo com duas linhas só.